File: distributed-rate-limiter/src/redis/lua/fixed_window_counter.lua
local key = KEYS[1]

local arr = redis.call('HGETALL', key)
-- 沒有限流，直接通過
if #arr == 0 then
    return 1
end

-- 將數組轉換為對象
local obj = {}
for i = 1, #arr, 2 do
    obj[arr[i]] = tonumber(arr[i + 1])
end

local current_time = tonumber(redis.call('TIME')[1])
if obj.startTime + obj.timeWindows <= current_time then
    -- 重置計數器和窗口起始時間
    redis.call('HSET', key, 'count', 1, 'startTime', current_time)
    return 1
end

if obj.count < obj.maxRequests then    
    redis.call('HSET', key, 'count', obj.count + 1)
    return 1
end

-- 超過限制，拒絕請求
return 0
    

File: distributed-rate-limiter/src/rate-limiter/fixed_window_counter.ts
import { RateLimiter, fixedWindowCounterConfig, fixedWindowCounterResponse } from './typeModel';
import * as _ from '../constants';
import { redisClient } from '../redis';

/**
 * 固定窗口計數器限流器實現
 * timeWindows: number; // 時間窗口大小 (秒)
 * maxRequests: number; // 最大請求數
 * count: number; // 當前請求數
 * startTime: number; // 窗口開始時間
 */
export default class FixedWindowCounter implements  RateLimiter {
    async run(key: string): Promise<fixedWindowCounterResponse> {
        return Boolean(await (redisClient as any)[_.FIXED_WINDOW_COUNTER](1, key, ''));
    }

    async set(key: string, config: fixedWindowCounterConfig): Promise<boolean> {
        return Boolean(await redisClient.hset(key, 
            'timeWindows', config.timeWindows, 
            'maxRequests', config.maxRequests, 
            'count', 0, 
            'startTime', Math.floor(Date.now() / 1000)));
    }

    async delete(key: string): Promise<boolean> {
        return Boolean(await redisClient.del(key));
    }
}